Check specific Visual Studio configurations

fuzzelhjb opened the pull request:
Add support to check against selected Visual Studio configurations in a cppcheck project file.
The cppcheck project file format was extended with a potential list of VS configurations. On importing a Visual Studio solution or project file from a .cppcheck file only those configurations that match the ones in the list are checked.

danmar commented on lib/importproject.h:
```diff
@@ -153,6 +154,8 @@ namespace CppcheckXml {
     const char MaxCtuDepthElementName[] = "max-ctu-depth";
     const char CheckUnknownFunctionReturn[] = "check-unknown-function-return-values";
     const char Name[] = "name";
+    const char VSConfigurationElementName[] = "vsconfigurations";
```
The variable names are fine. But the xml elements normally use "-" to separate names. So maybe "vs-configurations" and "config" are good names.

fuzzelhjb commented:
Ok I'll update the names.

Daniel Marjamäki <notifications@github.com> schrieb am Sa., 25. Jan. 2020,
10:11:

> *@danmar* commented on this pull request.
>
> thanks. this is probably 95% good.
>
> I think I need to build and test it..
> ------------------------------
>
> In lib/importproject.h
> <https://github.com/danmar/cppcheck/pull/2503#discussion_r370923187>:
>
> > @@ -153,6 +154,8 @@ namespace CppcheckXml {
>      const char MaxCtuDepthElementName[] = "max-ctu-depth";
>      const char CheckUnknownFunctionReturn[] = "check-unknown-function-return-values";
>      const char Name[] = "name";
> +    const char VSConfigurationElementName[] = "vsconfigurations";
>
> The variable names are fine. But the xml elements normally use "-" to
> separate names. So maybe "vs-configurations" and "config" are good names.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2503?email_source=notifications&email_token=AAQTYQ6JCPEIBC3BGPWRCM3Q7P64ZA5CNFSM4KLFXBNKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCTBOQLA#pullrequestreview-348317740>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQY4KFJ4PLZSF7NIH43Q7P64ZANCNFSM4KLFXBNA>
> .
>

danmar commented:
I assume this technically is working but maybe the UI can be tweaked.

For me it is confusing what happens if I both click on "Analyze all visual studio solutions" and write something in the edit box. If "analyze all visual studio solutions" is checked I suggest that the editbox is disabled.

Somehow it would be good if the available configurations is listed. Do you want to be able to analyze several configurations but not all? If you are fine analyzing 1 configuration then a combobox could be used.

fuzzelhjb commented:
I changed the behavior as suggested. If you select "Analyze all visual studio configurations" then the textbox is disabled. I think it would be nice to define multiple configurations so a semicolon separated list would make more sense here.
Afaik the UI doesn't read the actual vcxproj/sln file so there is no way to list all configurations in the textbox.

danmar commented:
> Afaik the UI doesn't read the actual vcxproj/sln file so there is no way to list all configurations in the textbox.

That is true. But that would be pretty simple.

    #include "importproject.h"

In `ProjectFileDialog::browseImportProject()`:

I believe it would be pretty simple to create a utility function like:

    static QStringList getProjectConfigs(QString fileName) {
        QStringList ret;
        ImportProject importer;
        importer.load(fileName.toStdString());
        for (const std::string &cfg : importer.getConfigs())
            ret << QString::fromStdString(cfg);
        return ret;
    }

And then in `ImportProject` we need a function `getConfigs()`.

danmar commented:
I am grateful that you are working on this. It will be a really cool enhancement!

fuzzelhjb commented:
You're welcome :) I updated the importer to automatically read all different configurations. The UI was updated too. I replaced the textbox with a QListWidget so the user can multi-select configurations

danmar commented on lib/importproject.cpp:
```diff
@@ -621,8 +621,16 @@ void ImportProject::importVcxproj(const std::string &filename, std::map<std::str
-                        if (p.platform != ProjectConfiguration::Unknown)
+                        if (p.platform != ProjectConfiguration::Unknown) {
                             projectConfigurationList.emplace_back(cfg);
+                            bool alreadyIn = false;
```
hmm.. I envisioned that this would be kept out of `importVcxproj`.

    std::set<std::string> ImportProject::getVSConfigs() const {
        std::set<std::string> ret;
        for (const FileSettings &fs: fileSettings) {
            ret.insert(fs.cfg);
        }
        return ret;
    }

I guess I don't have a strong opinion, I could allow that we have it in `importVcxproj`... but at least use `std::set` instead so you don't need the extra code to avoid duplicates.

danmar commented:
> The UI was updated too. I replaced the textbox with a QListWidget so the user can multi-select configurations

Cool.. will take a look soon

danmar commented:
ok.. after I merged the clang-tidy pull request there is a conflict. Can you please update this pull request? First you update your master branch so it is synced with my master. Then hopefully you just need these 3 commands:

    git checkout check_specific_VS_config
    git rebase master
    git push -f

danmar commented:
I merged this, it works good enough for merge. But I think you can tweak it better.
 * When I did not have a vcxproj/sln and chose a sln, both the checkbox and the listbox was enabled.
 * When I remove the vcxproj/sln from the editbox then the list box is not cleared.

fuzzelhjb commented:
I created a new PR with some tiny fixed that should fix this.
